File: web/src/types/knowledge.ts
export type KnowledgeCommand = {
  command: string
  purpose: string
  outcome: string
}

export type KnowledgeAssetRef = {
  assetId: number | null
  label: string
}

export type KnowledgeSourceRef = {
  conversationId: string | null
  eventId: string | null
  eventIndex: number | null
  eventType: string
  quote: string
  relevance: string
}

export type KnowledgeSourceConversation = {
  id: string | null
  title: string
  updatedAt: string | null
}

export type KnowledgeDraft = {
  title: string
  summary: string
  problem: string
  diagnosis: string
  resolution: string
  commands: KnowledgeCommand[]
  assets: KnowledgeAssetRef[]
  tags: string[]
  sources: KnowledgeSourceRef[]
  redactionWarnings: string[]
}

export type KnowledgeEntry = Omit<KnowledgeDraft, 'redactionWarnings'> & {
  id: string
  sourceConversation: KnowledgeSourceConversation
  createdAt: string
  updatedAt: string
}

export type KnowledgeSearchResponse = {
  items: KnowledgeEntry[]
  total: number
  limit: number
  offset: number
}

export type KnowledgeReindexResponse = {
  indexed: number
  failed: number
}

export type KnowledgeGenerateDraftResponse = {
  draft: KnowledgeDraft
  sourceConversation: KnowledgeSourceConversation
}

export type KnowledgeEntryPayload = KnowledgeDraft & {
  sourceConversationId?: string | null
  sourceConversationTitle?: string
  sourceConversationUpdatedAt?: string | null
}

export type KnowledgeSearchParams = {
  query?: string
  assetId?: number
  tag?: string
  sourceConversationId?: string
  limit?: number
  offset?: number
}

File: src/app/api/schemas/knowledge.py
from pydantic import BaseModel, Field


class KnowledgeCommandView(BaseModel):
    command: str = ""
    purpose: str = ""
    outcome: str = ""


class KnowledgeAssetRefView(BaseModel):
    assetId: int | None = None
    label: str = ""


class KnowledgeSourceRefView(BaseModel):
    conversationId: str | None = None
    eventId: str | None = None
    eventIndex: int | None = None
    eventType: str = ""
    quote: str = ""
    relevance: str = ""


class KnowledgeSourceConversationView(BaseModel):
    id: str | None = None
    title: str = ""
    updatedAt: str | None = None


class KnowledgeDraftView(BaseModel):
    title: str = ""
    summary: str = ""
    problem: str = ""
    diagnosis: str = ""
    resolution: str = ""
    commands: list[KnowledgeCommandView] = Field(default_factory=list)
    assets: list[KnowledgeAssetRefView] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    sources: list[KnowledgeSourceRefView] = Field(default_factory=list)
    redactionWarnings: list[str] = Field(default_factory=list)


class KnowledgeGenerateDraftRequest(BaseModel):
    maxSourceEvents: int = 120
    modelName: str | None = None


class KnowledgeGenerateDraftResponse(BaseModel):
    draft: KnowledgeDraftView
    sourceConversation: KnowledgeSourceConversationView


class KnowledgeEntryCreateRequest(BaseModel):
    title: str = ""
    summary: str = ""
    problem: str = ""
    diagnosis: str = ""
    resolution: str = ""
    commands: list[KnowledgeCommandView] = Field(default_factory=list)
    assets: list[KnowledgeAssetRefView] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    sources: list[KnowledgeSourceRefView] = Field(default_factory=list)
    redactionWarnings: list[str] = Field(default_factory=list)
    sourceConversationId: str | None = None
    sourceConversationTitle: str = ""
    sourceConversationUpdatedAt: str | None = None


class KnowledgeEntryUpdateRequest(BaseModel):
    title: str = ""
    summary: str = ""
    problem: str = ""
    diagnosis: str = ""
    resolution: str = ""
    commands: list[KnowledgeCommandView] = Field(default_factory=list)
    assets: list[KnowledgeAssetRefView] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    sources: list[KnowledgeSourceRefView] = Field(default_factory=list)
    redactionWarnings: list[str] = Field(default_factory=list)
    sourceConversationId: str | None = None
    sourceConversationTitle: str = ""
    sourceConversationUpdatedAt: str | None = None


class KnowledgeEntryView(BaseModel):
    id: str
    title: str
    summary: str = ""
    problem: str = ""
    diagnosis: str = ""
    resolution: str = ""
    commands: list[KnowledgeCommandView] = Field(default_factory=list)
    assets: list[KnowledgeAssetRefView] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    sources: list[KnowledgeSourceRefView] = Field(default_factory=list)
    sourceConversation: KnowledgeSourceConversationView = Field(default_factory=KnowledgeSourceConversationView)
    createdAt: str
    updatedAt: str


class KnowledgeSearchResponse(BaseModel):
    items: list[KnowledgeEntryView] = Field(default_factory=list)
    total: int = 0
    limit: int = 20
    offset: int = 0


class KnowledgeReindexResponse(BaseModel):
    indexed: int = 0
    failed: int = 0
